src/borg/key.py

```diff
@@ -154,13 +156,82 @@ class AESKeyBase(KeyBase):
 
     PAYLOAD_OVERHEAD = 1 + 32 + 8  # TYPE + HMAC + NONCE
 
+    def __init__(self, repository):
+        super(AESKeyBase, self).__init__(repository)
+        self.key_nonce_file = os.path.join(get_key_nonces_dir(), self.repository.id_str)
+
     def id_hash(self, data):
         """Return HMAC hash using the "id" HMAC key
         """
         return hmac_sha256(self.id_key, data)
 
+    def get_local_free_nonce(self):
+        if os.path.exists(self.key_nonce_file):
+            with open(self.key_nonce_file, 'r') as fd:
+                return bytes_to_long(unhexlify(fd.read()))
+        else:
+            return None
+
+    def commit_local_nonce_reservation(self, next_unreserved, start_nonce):
+        if self.get_local_free_nonce() != start_nonce:
+            raise Error("nonce space reservation with mismatched previous state")
+        with SaveFile(self.key_nonce_file, binary=False) as fd:
+            fd.write(bin_to_hex(long_to_bytes(next_unreserved)))
+
+    def get_repo_free_nonce(self):
+        try:
+            repo_free_nonce = self.repository.get_free_nonce()
+        except InvalidRPCMethod as error:
+            # old server version, suppress further calls
+            self.get_repo_free_nonce = lambda: None
+            self.save_repo_end_of_nonce_reservation = lambda next_unreserved, start_nonce: None
+            return None
+        if repo_free_nonce is not None:
+            repo_free_nonce = bytes_to_long(repo_free_nonce)
```

Comment:
I prefer to use int, as all used need to do arithmetic and having the call sites do the conversion while handling None is messy. Unless you see some problem with sending uint64 sized values as int via msgpack.